### DISCUSSION TITLE
Figure out some process

### DISCUSSION THREAD

**Discussion by fstamour** [fstamour/cl-car#6] [General]
- no need to be complicated
- no need to be final

**Comment by Gleefre** [comment]
Suggestion:

-- Each CDR idea should get a separate issue for a discussion. There should be several tags for issues, for example:
`idea` - just an idea, time for discussions & writing a draft is ahead;
`draft` - a draft was written, time for discussions before sending to CDR;
`sent` - draft sent to CDR, all discussions should happen there now;
`rejected` - idea/draft rejected (issue closed with summary reason why this idea was rejected)
`final` - this idea was accepted & finalized by CDR.


--? It might be useful to have some sort of pinned message in an issue that helps navigating the discussion if it becomes to large. Dunno if this is possible. (Maybe the first comment should be reserved for that?)


-- All ideas should get a unique number identifier based on the first/main issue's number.


-- All drafts / changes to them should be submitted via PRs; new drafts should be created in the folder `drafts/idea-x/...`. (?) Maybe submodules could be supported? Dunno.


-- Each draft should have a separate README file. It should include a link to the CDR's issue; link to the CDR document (when sent) or reason why it was rejected/cancelled (if it was rejected or cancelled); author(s) information; license information; information on how to read/build the document when applicable, a link to an html page if applicable.
(?) Each draft should have a LICENSE file, with some sort of permissive license used by default. Something MIT-like?.. Dunno. It should not be restrictive and it should exist so that we don't have a "HyperSpec licensing situation"


-- There should be a description of current behavior (on various implementations) in a draft when applicable (for example PLNs).


-- If there are "Issues" for a specific draft, they should be located in `drafts/idea-x/issues/...` folder, one file (`issue-x.<ext>`) for each issue.

This file should describe this issue, and provide current behavior (on various implementations) when applicable.

An issue could get a separate global issue to be discussed in; or it can be discussed on this idea's main issue. Alternatively it could be discussed on a *thread* in the *github discussion* corresponding to this cdr idea.


-- There should be a pinned issue / `status.org` file with list of possible CDRs and their status. This would make it easier to see if there already exists a CDR for a given idea.

For example (here checking the checkbox would mean that this CDR passed the given stage and it is time to move the entry onto the next stage):

1. Stage: idea
- [x] Structures (issue #x)
- [ ] Package Locks (issue #y)
2. Stage: draft written.
- [ ] PLNs (issue #z)
3. Stage: draft sent to CDR
- [ ] Quux (issue #t)
4. Stage: Implementation Support
- [ ] FooBar (issue #z) (cdr-N)  - FOOCL, BARCL
5. Stage: rejected
- Quix (issue #l)

**Reply by fstamour** [reply]
It's hard to answer to this message, there's so stuff in it

**Reply by fstamour** [reply]
The "checkboxes to track the maturity" could go in an issue template 

Is there a "discussion template"?

Yes! Kinda: https://docs.github.com/en/discussions/managing-discussions-for-your-community/creating-discussion-category-forms

**Reply by Gleefre** [reply]
Hm, but those are not part of the git repo. (Checkboxes in the example up there are not really needed; but sorting & grouping list of existing IFCDRs and CDDRs by their maturity would be nice IMO.)

**Reply by Gleefre** [reply]
Checkboxes could go in an issue/discussion template, right -- those would make much more sense there.

**Comment by fstamour** [comment]
I didn't read your comments yet, I just came here to say:

I think the important stuff should be kept in the git repository itself, so that it isn't tied to github

**Comment by Gleefre** [comment]
Hm, agree on that, but what exactly do you think should be important?

Obviously drafts are important, I guess the "current status of all ideas" file would be important too. Are concrete discussions important? Keeping those as files probably is not convenient.

Maybe some sort of "logging" could be applied, like `conversation.txt` file for an idea that summarizes the discussion (proposals, opinions, maybe quotes, -- similar to X3J13 Cleanup Issues)

**Reply by fstamour** [reply]
"current status of all ideas" files could simply be "readme.org"

**Reply by fstamour** [reply]
At the root of the repository

**Reply by fstamour** [reply]
A bit like what I do here, but more formal: https://gitlab.com/fstamour/catchall/-/blob/2bf4679ba31919ce0b5d70c861a3bd6811197f67/readme.org

**Reply by Gleefre** [reply]
Right, it can be located in the README

**Comment by Gleefre** [comment]
One more thought on the process: some CDRs could & should have a "test system" being written. These are important as well and could be stored in `test-systems/idea-<x>/...`. One thing to keep in mind is that these probably should be git submodules, so that it is possible to submit those to quicklisp (or maybe to create & host a separate quicklisp distribution later, much much later)

**Comment by fstamour** [comment]
"concrete discussions" are important, but they can be heavily edited to make them more useful

Have you ever read the "issues" in the hypespec (though, I find them often a _too_ terse)

**Comment by fstamour** [comment]
1 issue per CDR sounds good

**Reply by fstamour** [reply]
Perhaps one discussion per CDR?

**Reply by Gleefre** [reply]
Makes sense, yes. Even though issues are much simpler to use & are... more minimalist?.. discussions seems to have threads which would make discussing CDRs much much easier.

**Reply by fstamour** [reply]
My thoughts exactly

**Comment by fstamour** [comment]
The tags sounds good too

**Reply by Gleefre** [reply]
Do we have tags for discussions?

**Reply by Gleefre** [reply]
Seems like those exist, yes.

**Reply by fstamour** [reply]
I created labels based vaguely on SRFI's process:

> When the proposal has been vetted by the editors, it receives its SRFI number and becomes draft. 

> [...] draft, final and withdrawn SRFIs

![Screenshot_20231128-000000_Firefox](https://github.com/fstamour/trivial-cdr/assets/2881922/75435fee-5b08-42cd-8133-51215eecca4f)

But it doesn't match exactly what this project would be for...

This project would be for stuff that is not even a draft yet

**Reply by Gleefre** [reply]
I'd say just add one more stage: `idea`?

**Comment by Gleefre** [comment]
> Have you ever read the "issues" in the hypespec (though, I find them often a _too_ terse)

My thoughts exactly :)

**Comment by fstamour** [comment]
Agree on having tests and CI

Not so sure about git submodules, that's pretty far from "keeping it simple" xD

**Comment by Gleefre** [comment]
Hm, I didn't really talk about CI though -- it could be hard to set up; and it is not entirely obvious what should it be testing.

By a test system I meant a system (asdf system) that you can load and run tests on to check if an implementations adheres to the standard defined in this CDR or not. They could also have another functionality -- define a `trivial-...` package that contains defined symbols.

That is, I was talking about something similar to `trivial-package-local-nicknames`.

**Reply by fstamour** [reply]
About CI: it's easy if you don't use GitHub's CI. That's why I'm slowly migrating to GitLab. I'm pretty comfortable with GitLab's CI... So I don't find this hard.

About test: every piece of code can ve tested, even if the test is just "read-file", that would catch syntax errors for example.

**Reply by Gleefre** [reply]
But these tests -- they would be more like `ansi-test` (https://gitlab.common-lisp.net/ansi-test/ansi-test/-/issues), but `cdr-tests`. Does it makes sense to run tests against any implementations? Maybe as an automated way to have "supported by" table for example, but dunno.

**Reply by fstamour** [reply]
I understand, what I mean is that you can have some (perhaps minimal) kind of checks _on these test systems_...

**Comment by Gleefre** [comment]
I'm not sure how exactly submodules work; but those would allow the author of a draft/test-system to work on it in its own repository. Also those would allow to easily add test-systems to quicklisp / ultralisp.

(A naming scheme other than `trivial-...` should be chosen though, as there are already trivial-... systems.)

**Comment by Gleefre** [comment]
About identifiers:

Ideas could be called `if-cdr-<name>`, as in "Idea For: Common lisp Document Repository"
Drafts could be called `cddr-<name>`, as in "Common Lisp Document Draft Repository"
(Drafts could also use it a placeholder for a feature name (see cdr-14) until it is sent to CDR)

(Should it go in #2 maybe?..)

**Comment by fstamour** [comment]
I'll try converting this issue into a discussion, so we can have threads

**Reply by Gleefre** [reply]
Nice, converting issue to a discussion is possible :).

**Comment by fstamour** [comment]
For "inspiration":

https://www.w3.org/2023/Process-20231103/

**Comment by fstamour** [comment]
https://srfi.schemers.org/srfi-process.html

**Reply by fstamour** [reply]
https://srfi.schemers.org/srfi-template.html

**Comment by fstamour** [comment]
@Gleefre give me some ideas to improve #23 I'm brain dead at the moment :sweat_smile:

**Reply by Gleefre** [reply]
Looks great!

A few possible ideas:
  - Link to it from the README or maybe instead copy/move it there.
  - Create a pinned discussion/issue or a section in the README with a list of all ideas with their stages.

    Now that there are more ideas (mostly from the gist of course), it would be much easier to scan a compact list of ideas instead of the list of discussions with the corresponding tag.

  - Mention the CDR site as another place to check for existing ideas.

  - Add directions for owners of similar lists.

    Like to open an issue similar to [#4](https://github.com/fstamour/cl-car/issues/4), [#12](https://github.com/fstamour/cl-car/issues/12) and [#15](https://github.com/fstamour/cl-car/issues/15). Maybe mention the possibility of migrating/copying ideas.

    _A few ideas out of the scope of the 'Welcome' page:_
    - _It would be nice to create tags for ideas from other lists._
    - _Probably need to contact owners of other lists to ask if they are okay with copying ideas here, if we want to do that at all._

**Reply by Gleefre** [reply]
One more idea:
- Add some sort of encouragement for participation in already existing discussions.

**Reply by fstamour** [reply]
All good ideas I think

I already planned on adding the ideas to the repo, as I mentioned somewhere...
